File: apps/ios/ExpCore/Sources/Domain/PlaceholderStatus.swift
import Foundation

// EXP-630 placeholder members: a member whose email invite is still
// unaccepted has not joined yet. `pending` = the link still works, `expired`
// = it lapsed or was revoked (the roster row stays until removed; resending
// mints a fresh link — a web-only surface, iOS only badges the row).
//
// EXP-1076 adds `unsent`: the Linear import seats everyone it found so
// attributions land, but nobody was ever asked to join — those rows carry
// `sent_at` NULL (and an already-lapsed `expires_at`, so expiry readers treat
// the token as dead). "Invite expired" would be a lie there; it reads "Not
// invited".
// Mirrors web `lib/placeholder-status.ts`.
public enum PlaceholderStatus: String, Sendable {
    case unsent
    case pending
    case expired

    /// The badge copy, byte-identical to web's `PLACEHOLDER_LABELS`.
    public var label: String {
        switch self {
        case .unsent: return "Not invited"
        case .pending: return "Invited"
        case .expired: return "Invite expired"
        }
    }

    /// Web's `RANK`: a member may hold several rows (a superseded link, then
    /// a fresh one) — the best news wins. A live link outranks a lapsed one,
    /// and any issued link outranks "never invited".
    fileprivate var rank: Int {
        switch self {
        case .pending: return 2
        case .expired: return 1
        case .unsent: return 0
        }
    }
}

/// The placeholder state per member id, read off the synced invites of one
/// team. An invite counts only while it is bound to a placeholder member AND
/// unaccepted; `sentAt` nil is `unsent` whatever the expiry says, else
/// `expiresAt` in the future is `pending`, else `expired`.
///
/// An unparseable `expires_at` reads as `expired` — the same fail-closed
/// choice web makes (an invalid `Date` never compares greater than `now`).
public func placeholderStatuses(
    invites: [TeamInviteEntity],
    now: Date = Date()
) -> [String: PlaceholderStatus] {
    var result: [String: PlaceholderStatus] = [:]
    for invite in invites {
        guard let placeholderUserId = invite.placeholderUserId,
              !placeholderUserId.isEmpty,
              invite.acceptedAt == nil
        else { continue }
        let status: PlaceholderStatus
        if invite.sentAt?.isEmpty ?? true {
            // Never sent ⇒ never expired: the import stamps
            // `expires_at = created_at` on purpose, so expiry says nothing
            // about this row.
            status = .unsent
        } else {
            let expiresAt = WireTimestamps.parse(invite.expiresAt)
            status = (expiresAt.map { $0 > now } ?? false) ? .pending : .expired
        }
        // pending > expired > unsent for the same member, in any input order.
        if let current = result[placeholderUserId], current.rank >= status.rank { continue }
        result[placeholderUserId] = status
    }
    return result
}

File: apps/web/src/lib/placeholder-status.ts
import type { TeamInvite } from "@/db/schema"

// EXP-630 placeholder members: a member whose email invite is still
// unaccepted has not joined yet. `pending` = the link still works, `expired`
// = it lapsed or was revoked (the roster row stays until removed; "Resend
// invite" mints a fresh link, at a corrected address if need be).
//
// EXP-1076 adds `unsent`: the Linear import seats everyone it found so
// attributions land, but nobody was ever asked to join — those rows carry
// `sent_at` NULL (and an already-lapsed `expires_at`, so expiry readers treat
// the token as dead). "Invite expired" would be a lie there; it reads "Not
// invited" and offers a first send.
export type PlaceholderStatus = `unsent` | `pending` | `expired`

// A member may hold several rows (a superseded link, then a fresh one): the
// best news wins — a live link outranks a lapsed one, and any issued link
// outranks "never invited".
const RANK: Record<PlaceholderStatus, number> = {
  pending: 2,
  expired: 1,
  unsent: 0,
}

export function placeholderStatuses(
  invites: Pick<
    TeamInvite,
    `placeholderUserId` | `acceptedAt` | `expiresAt` | `sentAt`
  >[],
  now: Date = new Date()
): Map<string, PlaceholderStatus> {
  const result = new Map<string, PlaceholderStatus>()
  for (const invite of invites) {
    if (!invite.placeholderUserId || invite.acceptedAt) continue
    // Never sent ⇒ never expired: the import stamps `expires_at = created_at`
    // on purpose, so expiry says nothing about this row.
    const status: PlaceholderStatus = !invite.sentAt
      ? `unsent`
      : new Date(invite.expiresAt) > now
        ? `pending`
        : `expired`
    const current = result.get(invite.placeholderUserId)
    if (current !== undefined && RANK[current] >= RANK[status]) continue
    result.set(invite.placeholderUserId, status)
  }
  return result
}

export const PLACEHOLDER_LABELS: Record<PlaceholderStatus, string> = {
  unsent: `Not invited`,
  pending: `Invited`,
  expired: `Invite expired`,
}
